### PR TITLE
Update Pseudocode_Final

### PR DESCRIPTION
**Pull request by patelami3431**
Tried to make it consistent with the documentation.md! Added some comments to make the working of the class better and feel free to add the review for my comments.

### CODE REVIEW

File: Documents/Pseudocode/Pseudocode_Final [hunk 1]
```diff
@@ -58,8 +59,8 @@ Class Network
 		vector<matrix> sum_nabla_b;
 		vector<matrix> sum_nabla_w;
 		vector<int> test_data_indices;
-		vector<int> mini_batch_indices;		//only used in train(), why not limit its scope to that function?
-		int num_correct;			//only used in train(), why not limit its scope to that function?
+		vector<int> mini_batch_indices;		//mini_batch_indices are being used in SGD() along with Train() [-Ami]
```
Comment (kjhsdgf) [inline Documents/Pseudocode/Pseudocode_Final:62]: But SGD() is only called in train(), so it's still within scope to place in train()

File: Documents/Pseudocode/Pseudocode_Final [hunk 2]
```diff
@@ -101,8 +102,9 @@ Class Network
 									//just make another data member for validation data
 		//open previous_network_filename
 		//open validation_data_filename
-		//readInit(previous_network_filename)
-		//classify(validation_data_filename)
+		//readInit(previous_network_filename)	//This method creates the vector of matrices for sum_nabla_b and other matrics that are not needed during classification, consider creating the vector matrices needed inside of this function [-Ami]
+							// parameter not needed for classify () as the training_data_filename stores the name of the validation file [-Ami]
+							//classify is already being called by main() [-Ami]
```
Comment (kjhsdgf) [inline Documents/Pseudocode/Pseudocode_Final:107]: weren't we allowing the Network constructor to be overriden such that passing it a previous network and a validation data file will automatically classify the data? If we're removing the classify() call, then we also have no need for the validation data file as a parameter.

File: Documents/Pseudocode/Pseudocode_Final [hunk 3]
```diff
@@ -254,17 +258,16 @@ Class Network
 	void classify() 				//we don't need any parameters here, because constructor stores the name of validation file in the training_data_file
 	{
 		//int ambiguous_data (0);
-		//int numTrainingInputs;		//will tell us the number of the training inputs in the verfification file
+		//int numTrainingInputs;		//will tell us the number of the training inputs in the verfification file	//consider numInputs [-Ami]
```
Comment (kjhsdgf) [inline Documents/Pseudocode/Pseudocode_Final:261]: That seems unclear, though numTrainingInputs is also vague in the same sense. Maybe something like numClassifications, because that's how we're using it in the code. Visual Basic auto-completes, so I don't think length of the name will ever be the problem. More or less, it will be problematic that the variable names don't align with what they do and/or what they represent in the code, so that they're unmemorable (when writing), or opaque (when reading).

tl;dr, I think whoever is writing the C++ code will probably be in the best position to label the variables within the scope of a function. I only change this stuff when I edit so I can keep track of what's going on as I write.

File: Documents/Pseudocode/Pseudocode_Final [hunk 4]
```diff
@@ -254,17 +258,16 @@ Class Network
...
 		//numTrainingInputs = size of file/size of a single verification datum
-		//Loop index i from 0 to size of (verification data) //from i = 0 to end
-			//call forwardPass(i), which takes a datum via training_data_filename and computes the activations <-- is this the right file?
+		//Loop index i from 0 to size of (verification data) //from i = 0 to numInput [-Ami]			//call forwardPass(i), which takes a datum via training_data_filename and computes the activations <-- is this the right file?
 			//set biggest = 0, which will act as the index of the biggest activation value
 			//Looping index j from 1 to size of (activations[num_layers]) (j=0 is assumed to be biggest)
 				//if activations(num_layers, j) > activations(num_layers, biggest) 
 					//then biggest = j
-				//else if activations(num_layers, j) == activations(num_layers, biggest)
+				//else if activations(num_layers, j) == activations(num_layers, biggest) 	//consider the ouput [0.1,0.1,0.2] [-Ami]
```
Comment (kjhsdgf) [inline Documents/Pseudocode/Pseudocode_Final:270]: Ay, muy correcto. I checked the history, but wouldn't the previous algorithm increment ambiguous_data for 
[0.1, 0.1, 0.1, 0.2]? 

Perhaps we could do an initial pass to assign the largest number, then a secondary to check whether or not that number appears twice:
//Looping index j from 1 to activations[num_layers].nc()
&nbsp;&nbsp;&nbsp;&nbsp;//if activations[num_layers](j, 1) > activations[num_layers](biggest, 1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//biggest = j
//Looping index j from 0 to activations[num_layers].nc()
&nbsp;&nbsp;&nbsp;&nbsp;//if activations[num_layers](j, 1) == activations[num_layers](biggest, 1)
&nbsp;&nbsp;&nbsp;&nbsp;//equalityCounter++
//If equalityCounter > 1
&nbsp;&nbsp;&nbsp;&nbsp;//increment ambiguous_data

Or use some flags:
//bool doubleCheck = false
//Looping index j from 1 to activations[num_layers].nc()
&nbsp;&nbsp;&nbsp;&nbsp;//if activations[num_layers](j, 1) > activations[num_layers](biggest, 1)
&nbsp;&nbsp;&nbsp;&nbsp;//biggest = j
&nbsp;&nbsp;&nbsp;&nbsp;//else if activations[num_layers](j, 1) == activations[num_layers](biggest, 1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//doubleCheck = true
//If doubleCheck
&nbsp;&nbsp;&nbsp;&nbsp;//bool isAmbiguous = false
&nbsp;&nbsp;&nbsp;&nbsp;//Looping index j from 1 to activations[num_layers].nc()
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//if activations[num_layers](i, 1) == activations[num_layers](biggest, 1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//isAmbiguous = true
&nbsp;&nbsp;&nbsp;&nbsp;//if isAmbiguous
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//ambiguous_data++

Anyway we end up doing it, it should work for arbitrarily-sized output layers.

**Review by kjhsdgf — CHANGES_REQUESTED** [review]
You're right about that classify loop, we need to fix it! See comments.

File: Documents/Pseudocode/Pseudocode_Final (hunk 2 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final:107]: to clarify: no need for validation data file because I considered the training file name to be written into the previous network file, viz. readInit(file)

File: Documents/Pseudocode/Pseudocode_Final (hunk 1 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:62]: No, if a variable is declared in train(), it can't be used in SGD(). This is because SGD() is an independent function.

File: Documents/Pseudocode/Pseudocode_Final [hunk 5]
```diff
@@ -58,8 +59,8 @@ Class Network
 		vector<matrix> sum_nabla_b;
 		vector<matrix> sum_nabla_w;
 		vector<int> test_data_indices;
-		vector<int> mini_batch_indices;		//only used in train(), why not limit its scope to that function?
-		int num_correct;			//only used in train(), why not limit its scope to that function?
+		vector<int> mini_batch_indices;		//mini_batch_indices are being used in SGD() along with Train() [-Ami]
+		int num_correct;			//we need this to be a class data member, because it is used in backPropagation() as well
```
Comment (patelami3431) [inline Documents/Pseudocode/Pseudocode_Final:63]: Yeah! So, my suggestion would be let that be the data member of the class for now.

File: Documents/Pseudocode/Pseudocode_Final (hunk 1 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final:62]: You know what, you're totally right, scope is limited to the brackets. Disregard these comments, then!

File: Documents/Pseudocode/Pseudocode_Final (hunk 2 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:107]: I was thinking about the fact that classification only requires the forwardPropagation(). So, we won't be needing the sum_nabla_b and sum_nabla_w matrices for classification. So, let's assign them them with a 0 (though, I am still doubtful about this!)

File: Documents/Pseudocode/Pseudocode_Final (hunk 3 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:261]: I see what you mean! How about using a name like num_batches?

File: Documents/Pseudocode/Pseudocode_Final (hunk 4 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:270]: I think we can think about this once we start coding! We will be getting more on how the code handles the cases like this.

File: Documents/Pseudocode/Pseudocode_Final (hunk 2 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:107]: Did you look at the previous one, where I made a constructor specifically for the classification? It wasn't calling readInit() for classification. readInit(string &) gets called only when we want to train an existing network.

Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final:107]: That's true, but we had a function that would do everything for us, so why not use it? Setup happens once, so it wouldn't result in any major slowdown.

Initializing the nabla vectors with zeros only happens at setup. The first time the user runs backPropagation() it will total the results with any values already present. The nabla vectors are initialized to 0 for the same reason they are zeroed out after every mini-batch, during SGD.

I had taken your meaning to override the constructor for classification to be: let the user give us a validation file and a previous (trained) network, then the network will produce output -- with only one line of code. And, at the time, I thought we wanted a constructor that took a filename to reconstruct an existing network, and the classification override in the previous version of pseudocode would have made that impossible, since the argument lists would have been the same (could be mistaken there). Additionally, having to run another constructor and train a whole network before using the classify constructor, and *then* run classify() just didn't seem to fit the description.

To my understanding, readinit(filename) gets called when you want to load/initialize a previous network from a file, regardless of whether you intend to train or classify.

File: Documents/Pseudocode/Pseudocode_Final (hunk 3 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final:261]: this is classify(), right? I don't think we use mini-batches, do we? I mean, for your own purposes, change names as you see fit. If I end up running through the code to update/troubleshoot, then I'll just change the names to whatever makes sense to me (if I need to).

File: Documents/Pseudocode/Pseudocode_Final (hunk 5 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final]: Sounds good to me!

File: Documents/Pseudocode/Pseudocode_Final (hunk 2 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final:107]: We could probably figure out ways to only load what we need, given the circumstances (i.e. which constructor we call), and you've pretty much got that covered. I would just make a minor request that for modularity's sake we defer the responsibility of initializing networks from files/cin to readInit() functions.

File: Documents/Pseudocode/Pseudocode_Final (hunk 3 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:261]: My suggestion would be using the name like endOfFile. I am sorry, I was mistaken. I thought that to be similar to SGD_Calls.

Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:261]: Or else, let's go with numData. That's what the variable is supposed to give us right? The number of data inputs to be classified from the given file.

File: Documents/Pseudocode/Pseudocode_Final (hunk 2 above)
Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:107]: I think we could do that, initializing the network matrices in readInit()! But, creating the matrices or initializing the parameters of the class usually happens in the constructor. readInit(), as the name suggests, is just responsible for reading in the values from the user.

Comment (patelami3431) [reply Documents/Pseudocode/Pseudocode_Final:107]: Actually, it doesn't matter where we create the matrices as such. So, I am okay with creating the required matrices in the readInit() methods.

File: Documents/Pseudocode/Pseudocode_Final (hunk 3 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final]: Either of these work!

File: Documents/Pseudocode/Pseudocode_Final (hunk 2 above)
Comment (kjhsdgf) [reply Documents/Pseudocode/Pseudocode_Final:107]: Sounds like a solid jurisdiction for readInit(): only to read contents of a file or cin. In practice, we might be able to pass a flag to signal to readInit() how much to take from the previous network file, or something. Otherwise, I think we can merge this branch after the edits we discussed are implemented.
